accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,279 @@
+multi-targeting scenarios this means you can only use preview features for the
+latest framework version.
+
+For platform-level preview features this restriction makes sense because it
+models reality. However, for library-level preview features not only does this
+restriction not make sense, it's highly undesirable. Library authors simply want
+a way to express "these parts of my API surface might change". This statement
+isn't tied to any runtime-, framework-, or language features.
+
+Another concern we had for platform-level preview features is that we wanted to
+ensure that application developers are aware when their applications have a
+dependency on them. That is, we want to avoid a case where a library on NuGet
+uses platform-level preview features as an implementation detail that the
+application consumer isn't aware of. So we made the opt-in mechanism viral: a
+library that turns on platform-level preview features requires its consumer to
+turn them on as well.
+
+Since library-level preview features cannot change how the .NET runtime itself
+operates, we don't have that same level of concern. In fact, we believe the
+viral nature is also undesirable.
+
+For the remainder of this proposal we'll refer to *library-level preview*
+features as *experimental APIs*.
+
+## Scenarios and User Experience
+
+### Shipping an experimental API in a stable package
+
+Martin builds a library that integrates with ASP.NET Core and provides features
+to make it easier to build services. The library is distributed as a NuGet
+package. The package is already in a version well past 1.0 and is considered
+stable.
+
+He ships the package often and doesn't want to maintain multiple flavors of the
+package, one marked as prerelease and one marked as stable. However, he wants to
+expose new APIs without making the promise that these APIs are stable. He
+designs the new APIs as if they would be stable, that is, he puts them in the
+namespace they belong and names the types and members without any naming
+convention that implies that they aren't stable yet. If customer feedback is
+positive, the API will become stable as-is, otherwise he'll make changes to
+address the feedback.
+
+In order to convey to the consumers that these APIs aren't stable yet, he puts
+the `Experimental` attribute on them:
+
+```C#
+namespace MartinsLibrary.FancyLogging
+{
+    public static class FancyLoggerExtensions
+    {
+        [Experimental("ML123", UrlFormat="https://martinslibrary.net/diagnostics/{0}")]
+        public static IServiceCollection AddFancyLogging(this IServiceCollections services)
+        {
+            // ...
+        }
+    }
+}
+```
+
+### Consuming an experimental API
+
+David builds a new version of Jabber.NET and consumes Martin's library. He very
+much wants to play with the new fancy logging Martin added. When David calls
+`AddFancyLogging` he gets a compilation error:
+
+> error ML123: FancyLoggerExtensions.AddFancyLogging() is not a stable API. In
```

Comment:
I assume we can't offer any guidance on the diagnostic ID format, since folks need to make their own attempts to avoid overlap with others, if they care.
assuming so, when we ultimately doc this, we should not consistently use ML\d+ as you do here.

Reply:
Hm, that's a good point. AFAIK diagnostic IDs are just strings. Maybe should have a convention like `Product.Feature`? Could be more readable than a numeric literal.

@dotnet/roslyn-analysis what do you think?

Reply:
If we were to introduce experimental APIs within the shared framework, I presume we'd use `SYSLIB` as our prefix there too, perhaps starting with `SYSLIB5000` or something.

Reply:
@KrzysztofCwalina [also thinks](https://github.com/dotnet/designs/pull/285#issuecomment-1409387130) this would be better

Reply:
Hey, could we just do magic with the API signature instead? You want to allow use of API X.Y? Then use X.Y as the ID.

> error CS1234-X.Y: This API is still in active development.

So CS1234 would be a stable error id defined by the C# compiler, and the X.Y part is the FQN of the API. So the compiler emits unique error codes for each experimental API. A customer can then suppress warnings for the specific methods.

Reply:
It needs to match tooling expectations. Those may be fairly lax eg

https://github.com/dotnet/msbuild/blob/7cfb36cb90d1c9cc34bc4e0910d0c9ef42ee47b6/src/Shared/CanonicalError.cs#L70

VS has its own logic for passing non MSBuild tools; it hopefully matches more or less.

Also to consider is the UX eg width of the code column in the Error List. Users will expect a short code.

A variation of your suggestion could be to use the hashcode of the signature but either way this would require compiler work.

I recommend the conventional scheme used by every tool today: library owner picks any prefix they like, and then increments their suffix.

Reply:
@geeknoid not sure that scheme is desirable. Imagine an experimental feature that is a new type + members on existing types. You want to mark the type plus those members as being the same feature. That scheme would force each instance to have separate IDs.